dali/python/nvidia/dali/plugin/pytorch/proxy/__init__.py

```diff
@@ -0,0 +1,329 @@
+            processed_outputs = []
+            for output in outputs:
+                tensor = output.as_tensor()
+                torch_dtype = to_torch_type[tensor.dtype]
+                if isinstance(tensor, TensorGPU):
+                    torch_device = torch.device("cuda", self.pipe.device_id)
+                else:
+                    torch_device = torch.device("cpu")
+                processed_output = torch.empty(
+                    tensor.shape(),
+                    dtype=torch_dtype,
+                    device=torch_device,
+                )
+                cuda_stream = (
+                    torch.cuda.current_stream(device=torch_device)
+                    if isinstance(tensor, TensorGPU)
+                    else None
+                )
+                feed_ndarray(tensor, processed_output, cuda_stream=cuda_stream)
+                processed_outputs.append(processed_output)
+            processed_outputs = tuple(processed_outputs)
+        else:
+            processed_outputs = tuple([torch.from_dlpack(output.as_tensor()) for output in outputs])
+        return (info, processed_outputs)
+
+    def get_outputs(self, pipe_out_ref: DALIPipelineOutputRef):
+        req_info = pipe_out_ref.info
+        req_outputs = None
+        # If the data was already read, just return it (and clear the cache entry)
+        if req_info in self.cache_outputs:
+            req_outputs = self.cache_outputs[req_info]
+            del self.cache_outputs[req_info]
+
+        else:
+            info = None
+            # If not the data we are looking for, store it and keep processing until we find it
+            while req_info != info:
+                info, processed_outputs = self.next_outputs()
+                if info == req_info:
+                    req_outputs = processed_outputs
+                else:
+                    self.cache_outputs[info] = processed_outputs
+        # Unpack single element tuples
+        if isinstance(req_outputs, tuple) and len(req_outputs) == 1:
+            req_outputs = req_outputs[0]
+        return req_outputs
+
+    def thread_fn(self):
+        """
+        Asynchronous DALI thread that gets iteration data from the queue and schedules it
+        for execution
+        """
+        self.pipe.build()  # just in case
+
+        while not self.thread_stop_event.is_set():
+            try:
+                torch.cuda.nvtx.range_push("send_q.get")
+                info, inputs = self.send_q.get(timeout=5)
+                torch.cuda.nvtx.range_pop()
+            except mp.TimeoutError:
+                continue
+            except Empty:
+                continue
+
+            torch.cuda.nvtx.range_push(f"feed_input {info}")
+            for idx, input_name in enumerate(self.input_names):
+                self.pipe.feed_input(input_name, inputs[idx])
+            torch.cuda.nvtx.range_pop()
+
+            torch.cuda.nvtx.range_push(f"schedule_run {info}")
+            self.pipe.schedule_run()
```

Comment:
This should probably be `pipe._run_once`.
Normally `feed_input` should be called `prefetch_queue_depth` times before the first call to `run`.
Also, with `_run_once` it won't be necessary to use blocking external source.